compiler/rustc_typeck/src/check/method/probe.rs

```diff
@@ -154,6 +154,42 @@ enum ProbeResult {
     Match,
 }
 
+/// When adjusting a receiver we often want to do one of
+///
+/// - Add a `&` (or `&mut`), converting the recevier from `T` to `&T` (or `&mut T`)
+/// - If the receiver has type `*mut T`, convert it to `*const T`
+///
+/// This type tells us which one to do.
+///
+/// Note that in principle we could do both at the same time. For example, when the receiver has
+/// type `T`, we could autoref it to `&T`, then convert to `*const T`. Or, when it has type `*mut
+/// T`, we could convert it to `*const T`, then autoref to `&*const T`. However, currently we do
+/// (at most) one of these. Either the type has `T` and we convert it to `&T` (or with `mut`), or
+/// it has type `*mut T` and we convert it to `*const T`.
```

Comment:
It may be worth adding a note that while we could do both at the same time, it would have a different meaning -- that is, converting to a `&T` first would create a safe reference, which asserts (among other things) that the reference is aligned, etc.

Reply:
I think what you're saying is, if we allowed both autoref'ing converting a ref/ptr to const ptr, the order we do those would matter. Obviously the type of the result will be different, other than that:

- Autoref, convert to const (`*const T`): The pointer will be aligned (I'm not sure if I get this part right, I don't know about reference semantics in detail), among other things (what other things? I'll ask on Zulip)
- Convert to const first, then autoref (`&*T`): the pointer may not be aligned.

Did I get it right?

I'll try to update the comment.

Reply:
That sounds roughly correct, yes. One nit:

> The pointer will be aligned 

It's not that the pointer will be aligned, so much as that it would be UB if the pointer were *not* aligned (i.e., it'd be incorrect code, and the compiler might misoptimize it).